actionpack/lib/action_controller/metal/rendering.rb

```diff
@@ -4,6 +4,17 @@ module Rendering
 
     RENDER_FORMATS_IN_PRIORITY = [:body, :text, :plain, :html]
 
+    module ClassMethods
+      # Documentation at ActionController::Renderer#render
+      delegate :render, to: :renderer
+
+      # Returns a renderer class (inherited from ActionController::Renderer)
+      # for the controller.
+      def renderer
+        @renderer ||= Renderer.for(self)
```

Comment:
The `||=` is a read-check-write race condition, and since this is essentially a global (since it is stored on the class) we have to add a lock.  Here's a test to demo:

```ruby
require 'action_pack'
require 'action_controller'
require 'active_support'
require 'concurrent/atomic/count_down_latch'

wait_latch = Concurrent::CountDownLatch.new 100
go_latch = Concurrent::CountDownLatch.new

t = 100.times.map { |i|
  Thread.new {
    wait_latch.count_down
    go_latch.wait
    renderer = ActionController::Base.renderer
    renderer.defaults[:foo] = i
    renderer
  }
}

wait_latch.wait
go_latch.count_down

renderers = t.map(&:value)
p renderers.map(&:object_id).uniq.length
p renderers.uniq.map { |r| r.defaults[:foo] }
p ActionController::Base.renderer.defaults[:foo]
```

On my machine using MRI (`ruby 2.3.0dev (2015-09-12 trunk 51832) [x86_64-darwin14]`) I'm seeing between 24 and 34 unique instances of the renderer set on the controller class.  Also you will not be able to predict the value of `ActionController::Base.renderer.defaults[:foo]`.

I'm adding a mutex for now, but I'll see if I can change this to instantiate the renderer when the inherited hook fires so that we can have lock-free reads.